server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1324,6 +1325,7 @@ public void openEngineAndRecoverFromTranslog() throws IOException {
         };
         innerOpenEngineAndTranslog();
         getEngine().recoverFromTranslog(translogRecoveryRunner, Long.MAX_VALUE);
+        advanceMaxSeqNoOfUpdatesOrDeletes(seqNoStats().getMaxSeqNo());
```

Comment:
Yeah, I understand your point, but there is a reason that I have to put this initialization after we recover from translog.

1. We don't restore LocalCheckpointTracker when opening an engine but only bootstrap local_checkpoint and max_seq_no from the safe commit.

2. If we replay translog with the max_seq_no_updates set to the max_seq_no. We may duplicate operations whose seq_no between the max_seq_no from the safe commit and the max_seq_no from translog.

We can use the "isRetry" flag of index operations to solve this, but I am not sure because I think `isRetry` is attached to `autoGeneratedIdTimestamp`.